### DISCUSSION TITLE
[Bug]: Unknown authentication strategy "openid"

### DISCUSSION THREAD

**Discussion by Donavan** [danny-avila/LibreChat#674] [Troubleshooting] [answered]
### Contact Details

snark@hey.com

### What happened?

I'm trying to enable openID.  After enabling it the app blows up when I try to use the login

### Steps to Reproduce

Fill in the OPENID_* env variable.s

### Relevant log output

```shell
LibreChat  | Error: Unknown authentication strategy "openid"
LibreChat  |     at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:193:39)
LibreChat  |     at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:370:7)
LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
LibreChat  |     at next (/app/node_modules/express/lib/router/route.js:144:13)
LibreChat  |     at Route.dispatch (/app/node_modules/express/lib/router/route.js:114:3)
LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
LibreChat  |     at /app/node_modules/express/lib/router/index.js:284:15
LibreChat  |     at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
LibreChat  |     at next (/app/node_modules/express/lib/router/index.js:280:10)
LibreChat  |     at Function.handle (/app/node_modules/express/lib/router/index.js:175:3)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Hi @Donavan thanks for looking into this.

OpenID works for me on multiple machines, so I've moved this to a #troubleshooting discussion

First, assure the env variables are set
```bash
ALLOW_REGISTRATION=true
ALLOW_SOCIAL_LOGIN=true

OPENID_CLIENT_ID=your-id
OPENID_CLIENT_SECRET=your-client-Secret
OPENID_ISSUER=issuer_url
OPENID_SESSION_SECRET=your-session-secret
OPENID_SCOPE="openid profile email"
OPENID_CALLBACK_URL=/oauth/openid/callback
```

If all are set to correct values, I recommend you update your app to latest main branch.

⚠️ This assumes you made no changes to the main branch and you have at least npm and git installed.
```bash
git fetch origin
git checkout main
git pull origin main
```

Then to update, you have 3 options: 
```bash
npm run update #will prompt user if updating docker or local installation
# OR
npm run update:local
npm run update:docker
```

**Reply by Donavan** [reply]
I had just stood this up from a fresh checkout but to be safe followed your directions.  I still get the exact same error.

This is set up on an AWS EC2 instance behind a load balancer to supply HTTPS.  I'm attempting to use cognito as my open ID provider.

I think my config is correct, at least the values are filled in with what I believe is correct.  The error doesn't give me much to go on config wise.

**Reply by Donavan** [reply]
For what it's worth here's the log bits.

> LibreChat  | Connected to MongoDB
> LibreChat  | [Migrate] Nothing to migrate
> LibreChat  | There are 47 messages in the database, 47 indexed
> LibreChat  | There are 9 convos in the database, 9 indexed
> LibreChat  | Warning: connect.session() MemoryStore is not
> LibreChat  | designed for a production environment, as it will leak
> LibreChat  | memory, and will not scale past a single process.
> LibreChat  | Server listening on all interface at port 3080. Use http://localhost:3080 to access it
> LibreChat  | OPError: expected 200 OK, got: 400 Bad Request
> LibreChat  |     at processResponse (/app/node_modules/openid-client/lib/helpers/process_response.js:41:11)
> LibreChat  |     at Issuer.discover (/app/node_modules/openid-client/lib/issuer.js:179:18)
> LibreChat  |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
> LibreChat  |   error: 'expected 200 OK, got: 400 Bad Request'
> LibreChat  | }
> LibreChat  | No 'auth.json' file (service account key) found in /api/data/ for PaLM models
> LibreChat  | Error: Unknown authentication strategy "openid"
> LibreChat  |     at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:193:39)
> LibreChat  |     at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:370:7)
> LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat  |     at next (/app/node_modules/express/lib/router/route.js:144:13)
> LibreChat  |     at Route.dispatch (/app/node_modules/express/lib/router/route.js:114:3)
> LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat  |     at /app/node_modules/express/lib/router/index.js:284:15
> LibreChat  |     at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
> LibreChat  |     at next (/app/node_modules/express/lib/router/index.js:280:10)
> LibreChat  |     at Function.handle (/app/node_modules/express/lib/router/index.js:175:3)

**Comment by danny-avila** [comment]
> I had just stood this up from a fresh checkout but to be safe followed your directions. I still get the exact same error.
> 
> This is set up on an AWS EC2 instance behind a load balancer to supply HTTPS. I'm attempting to use cognito as my open ID provider.
> 
> I think my config is correct, at least the values are filled in with what I believe is correct. The error doesn't give me much to go on config wise.

The EC2/load balancer setup may be the culprit, but based on your errors maybe not. Possibly related discussion: https://github.com/danny-avila/LibreChat/discussions/471

My second thought is that if any of these variables get coerced to "falsy" values (empty string, undefined, etc.), openidStrategy will never be imported and that would make sense why it's "unknown"

```js
/* api\server\index.js */
  if (
    process.env.OPENID_CLIENT_ID &&
    process.env.OPENID_CLIENT_SECRET &&
    process.env.OPENID_ISSUER &&
    process.env.OPENID_SCOPE &&
    process.env.OPENID_SESSION_SECRET
  ) {
    app.use(
      session({
        secret: process.env.OPENID_SESSION_SECRET,
        resave: false,
        saveUninitialized: false,
      }),
    );
    app.use(passport.session());
    require('../strategies/openidStrategy');
  }
  ```

Before the above code, it might be useful to do a quick console.log
```js
  console.log(
    'is OpenID enabled?',!!(
      process.env.OPENID_CLIENT_ID &&
      process.env.OPENID_CLIENT_SECRET &&
      process.env.OPENID_ISSUER &&
      process.env.OPENID_SCOPE &&
      process.env.OPENID_SESSION_SECRET
    ),
  );
  console.dir({
    clientId: process.env.OPENID_CLIENT_ID,
    clientSecret: process.env.OPENID_CLIENT_SECRET,
    issuer: process.env.OPENID_ISSUER,
    scope: process.env.OPENID_SCOPE,
    sessionSecret: process.env.OPENID_SESSION_SECRET,
  }, { depth: null });
```

**Reply by danny-avila** [reply]
for solidarity, here's what I get 
```js
is OpenID enabled? true
{
  clientId: 'redacted',
  clientSecret: 'redacted',
  issuer: 'https://login.microsoftonline.com/redacted',
  scope: 'openid profile email',
  sessionSecret: 'secret'
}
```

**Reply by Donavan** [reply]
I get "Uncaught ReferenceError: process is not defined" in Firefox and Edge  trying your console log.  Standard login works fine through the load balancer so far.

**Reply by danny-avila** [reply]
you added it to this file right? 
```js
/* file path: api\server\index.js */
// insert console logs i provided here
  if ( // line 57
    process.env.OPENID_CLIENT_ID &&
    process.env.OPENID_CLIENT_SECRET &&
    process.env.OPENID_ISSUER &&
    process.env.OPENID_SCOPE &&
    process.env.OPENID_SESSION_SECRET
  ) {
  ```

**Reply by Donavan** [reply]
There we go:

> is OpenID enabled? true
> LibreChat  | {
> LibreChat  |   clientId: 'redacted',
> LibreChat  |   clientSecret: 'redacted',
> LibreChat  |   issuer: 'https://redacted.auth.us-east-2.amazoncognito.com/',
> LibreChat  |   scope: 'openid profile email',
> LibreChat  |   sessionSecret: 'redacted'
> LibreChat  | }

**Reply by danny-avila** [reply]
ok I think I see what's going. there may have been bad practice set as a precedent when the social logins were implemented, and they each followed the suite. Going to fix this and maybe this will still stop the race condition behavior you're seeing

**Reply by danny-avila** [reply]
@Donavan I pushed an update to the main branch that might potentially resolve your issue. you will have to remove the changes you made 

You can use `git checkout main`, `git reset --hard HEAD` before updating, that should put everything in sync

then `npm run update`

**Reply by Donavan** [reply]
Same error it seems:

> LibreChat  | Connected to MongoDB
> LibreChat  | [Migrate] Nothing to migrate
> LibreChat  | There are 5 messages in the database, 5 indexed
> LibreChat  | There are 1 convos in the database, 1 indexed
> LibreChat  | Warning: connect.session() MemoryStore is not
> LibreChat  | designed for a production environment, as it will leak
> LibreChat  | memory, and will not scale past a single process.
> LibreChat  | OPError: expected 200 OK, got: 400 Bad Request
> LibreChat  |     at processResponse (/app/node_modules/openid-client/lib/helpers/process_response.js:41:11)
> LibreChat  |     at Issuer.discover (/app/node_modules/openid-client/lib/issuer.js:179:18)
> LibreChat  |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> LibreChat  |     at async setupOpenId (/app/api/strategies/openidStrategy.js:41:20)
> LibreChat  |     at async /app/api/server/index.js:81:5 {
> LibreChat  |   error: 'expected 200 OK, got: 400 Bad Request'
> LibreChat  | }
> LibreChat  | Server listening on all interface at port 3080. Use http://localhost:3080 to access it
> LibreChat  | Error: Unknown authentication strategy "openid"
> LibreChat  |     at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:193:39)
> LibreChat  |     at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:370:7)
> LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat  |     at next (/app/node_modules/express/lib/router/route.js:144:13)
> LibreChat  |     at Route.dispatch (/app/node_modules/express/lib/router/route.js:114:3)
> LibreChat  |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
> LibreChat  |     at /app/node_modules/express/lib/router/index.js:284:15
> LibreChat  |     at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
> LibreChat  |     at next (/app/node_modules/express/lib/router/index.js:280:10)
> LibreChat  |     at Function.handle (/app/node_modules/express/lib/router/index.js:175:3)

**Reply by Donavan** [reply]
Ignore the above I know the cause of it.  I've gotten past that and I'm debugging another issue.

**Reply by danny-avila** [reply]
> Ignore the above I know the cause of it. I've gotten past that and I'm debugging another issue.

Awesome, mind sharing what you found?

**Reply by Donavan** [reply]
I have the issue fixed.  PR will be coming as soon as I add to the docs.

**Reply by Donavan** [reply, marked as answer]
#686

**Reply by Donavan** [reply]
Oh also @danny-avila :  If the OPENID_ISSUER is bad it will fail discovery and if it fails discovery you get the error I did initially.

**Reply by totopopov** [reply]
Guys is this working currently? 

I am getting the same logs using AzureID :
LibreChat         | Error: Unknown authentication strategy "openid"
LibreChat         |     at attempt (/app/node_modules/passport/lib/middleware/authenticate.js:193:39)
LibreChat         |     at authenticate (/app/node_modules/passport/lib/middleware/authenticate.js:370:7)
LibreChat         |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
LibreChat         |     at next (/app/node_modules/express/lib/router/route.js:149:13)
LibreChat         |     at Route.dispatch (/app/node_modules/express/lib/router/route.js:119:3)
LibreChat         |     at Layer.handle [as handle_request] (/app/node_modules/express/lib/router/layer.js:95:5)
LibreChat         |     at /app/node_modules/express/lib/router/index.js:284:15
LibreChat         |     at Function.process_params (/app/node_modules/express/lib/router/index.js:346:12)
LibreChat         |     at next (/app/node_modules/express/lib/router/index.js:280:10)
LibreChat         |     at /app/node_modules/express-rate-limit/dist/index.cjs:659:7

**Reply by totopopov** [reply]
Hey, my bad, all looks good, as specified in https://docs.librechat.ai/install/configuration/OAuth2-and-OIDC/azure.html, the  OPENID_ISSUER should be https://login.microsoftonline.com/Your Directory (tenant ID)/v2.0/ not https://login.microsoftonline.com/Your Directory (tenant ID)/oath/v2.0/

**Reply by danny-avila** [reply]
@totopopov great, glad you figured it out, thanks for letting me know.
